core/shared/src/main/scala/zio/ZIOMetric.scala

```diff
@@ -278,15 +300,15 @@ object ZIOMetric {
    * values over time.
    */
   abstract class Gauge[A](final val name: String, final val tags: Chunk[MetricLabel]) extends ZIOMetric[A] { self =>
-    private[this] val gauge = internal.metrics.Gauge(name, tags)
+    private[this] val gauge = ZFiberRef.unsafeMake(internal.metrics.Gauge(name, tags))
```

Comment:
The addition of a `FiberRef` wrapper here will impact performance. It may be better to make tags mutable if we can guard access sufficiently as to not violate referential transparency.

Reply:
One idea that would only make the construction non-threadsafe which I think is OK: 
- keep the non-fiber ref internal metric fields as they were but change them to `var`
- add a fiber-ref version too which is also a `var` and initialized to `null`
- the `taggedWith` sets the direct reference to `null` and initializes the fiber ref
- using a helper function to get the gauge from either the fiber ref or the direct ref to work with the metric

I was thinking a lot on how else could I achieve this - to be able to use the existing (and arbitray) aspects with dynamic tags but could not come up with anything that is better or more performant so if you have a better idea please let me know

Reply:
(by the way I did not do what I described here first because I assumed that `FiberRef` itself does something like this - that it's cheap until its never used on multiple fibers)